### PR TITLE
Final definite assignment

### PR DESCRIPTION
**Pull request by a14n**
## Description

Apply `final` on definite assignments.

## Related Issues

* dart-lang/sdk#40146

## Tests

I added the following tests:

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by goderbauer** [comment]
Looks like the analyzer is unhappy.

**Comment by a14n** [comment]
> Looks like the analyzer is unhappy.

This is what happens when you forget to save all files :facepalm:

**Review by goderbauer — APPROVED** [review]
I did not carefully examine every change, but what I looked at looked reasonable:

RSLGTM

Presumably, there's a lint for this in the works?

**Comment by a14n** [comment]
> Presumably, there's a lint for this in the works?

`prefer_final_locals` will catch them in the future but it requires some additions to analyzer.

File: packages/flutter/lib/src/animation/curves.dart [hunk 1]
```diff
@@ -738,9 +738,9 @@ class CatmullRomSpline extends Curve2D {
   Offset transformInternal(double t) {
     _initializeIfNeeded();
     final double length = _cubicSegments.length.toDouble();
-    double position;
-    double localT;
-    int index;
+    final double position;
```
Comment (whhone) [inline packages/flutter/lib/src/animation/curves.dart:741]: `position` is changed in line 745 (4 lines below). How can we make this variable `final`?

Comment (a14n) [reply packages/flutter/lib/src/animation/curves.dart:741]: > `position` is changed in line 745

`position` is rather initialized at line 745. With null-safety you can make a final declaration (the variable will not be initialized at declaration site) and initialize it later. The compiler/analyzer will ensure that when the variable is used it has been fully initialized. So you can have the following initialization:

```dart
final String s;
if (boolCondition) {
  s = 'a';
} else {
  s = 'b';
}
print(s); // if else branch is omitted you will get an error
```

Comment (whhone) [reply packages/flutter/lib/src/animation/curves.dart:741]: ah ha! Thanks for pointing this out. I was not aware of the new "null-safety" support in Dart.

**Pull request merged**